mitodl/mit-open
# Youtube video ETL and search

### What are the relevant tickets?
Closes #78 

### Description (What does it do?)
- Ports most of the Video ETL functionality from [open-discussions](https://github.com/mitodl/open-discussions/blob/master/course_catalog/etl/youtube.py) to mit-open. 
- Instead of crawling a github repo for various youtube channels each in their own yaml file, uses a single yaml file accessible via an http request (which can still be, but doesn't have to be, a github file).
- Makes both Videos and VideoPlaylists searchable learning resources (but not VideoChannels)
- The `pytube` package, used for transcripts in open-discussions, is [buggy](https://github.com/pytube/pytube/issues) even with the latest version (which means that the functionality is likely broken on open-discussions right now) and was replaced with the `youtube-transcripts-api` package , which is easier to use anyway.
- Does not yet port over [the functionality to assign topics](https://github.com/mitodl/open-discussions/pull/2424) to Videos and VideoPlaylists.  That will be handled in a subsequent PR.

Related ol-infrastructure PR: https://github.com/mitodl/ol-infrastructure/pull/2241

### How can this be tested?
- Set the following in your .env file:
  ```
  YOUTUBE_CONFIG_URL=https://raw.githubusercontent.com/mitodl/open-video-data/mb/one_file/youtube/channels.yaml
  YOUTUBE_DEVELOPER_KEY=<same value as open-discussions RC>
  ```
- Run `./manage.py backpopulate_youtube_data fetch`, when it finishes you should have a bunch of video and video playlist resources.
- Run `./manage.py backpopulate_youtube_data transcripts`, which will retrieve transcripts for the above videos.  
- Use the database and opensearch apis to filter by `resource_type=video` or `resource_type=video_playlist`, you should get results for each and the serialized data should look reasonable.
- Run a search query on some transcript text: `http://localhost:8063/api/v1/learning_resources_search/?resource_type=video&q="<some transcript text>"` - the video should be returned in results.
- Change `YOUTUBE_CONFIG_URL`  to https://raw.githubusercontent.com/mitodl/open-video-data/mb/one_file_no_mitx/youtube/channels.yaml, which omits an mitx channel.  Then rerun `./manage.py backpopulate_youtube_data fetch`.  After it is complete, the videos and playlists from that removed channel should no longer be published or appearing in search results.
